src/modules/manual_control/ManualControlSelector.cpp

```diff
@@ -0,0 +1,121 @@
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "ManualControlSelector.hpp"
+
+
+namespace manual_control
+{
+
+void ManualControlSelector::update_time_only(uint64_t now)
+{
+	if (_setpoint.chosen_input.timestamp_sample + _timeout < now) {
+		_setpoint.valid = false;
+		_instance = -1;
+	}
+}
+
+void ManualControlSelector::update_manual_control_input(uint64_t now, const manual_control_input_s &input, int instance)
+{
+	// This method requires the current timestamp explicitly in order to prevent weird cases
+	// if the timestamp_sample of some source is invalid/wrong.
+
+	// If previous setpoint is timed out, set it invalid, so it can get replaced below.
+	update_time_only(now);
+
+	if (_rc_in_mode == 0 && input.data_source == manual_control_input_s::SOURCE_RC) {
+		_setpoint = setpoint_from_input(input);
+		_setpoint.valid = true;
+		_instance = instance;
+
+	} else if (_rc_in_mode == 1 && (input.data_source == manual_control_input_s::SOURCE_MAVLINK_0
+					|| input.data_source == manual_control_input_s::SOURCE_MAVLINK_1
+					|| input.data_source == manual_control_input_s::SOURCE_MAVLINK_2
+					|| input.data_source == manual_control_input_s::SOURCE_MAVLINK_3
+					|| input.data_source == manual_control_input_s::SOURCE_MAVLINK_4
+					|| input.data_source == manual_control_input_s::SOURCE_MAVLINK_5)) {
+
+		// We only stick to the first discovered mavlink channel.
+		if (_setpoint.chosen_input.data_source == input.data_source || !_setpoint.valid) {
+			_setpoint = setpoint_from_input(input);
+			_setpoint.valid = true;
+			_instance = instance;
+		}
+
+	} else if (_rc_in_mode == 2) {
+		// FIXME: what to do in the legacy case?
+	} else if (_rc_in_mode == 3) {
+
+		// We only stick to the first discovered mavlink channel.
+		if (_setpoint.chosen_input.data_source == input.data_source || !_setpoint.valid) {
+			_setpoint = setpoint_from_input(input);
+			_setpoint.valid = true;
+			_instance = instance;
+		}
+
+	} else {
+		// FIXME: param value unknown, what to do?
+	}
+}
+
+manual_control_setpoint_s ManualControlSelector::setpoint_from_input(const manual_control_input_s &input)
+{
+	manual_control_setpoint_s setpoint;
+	setpoint.chosen_input.timestamp_sample = input.timestamp_sample;
+	setpoint.chosen_input.x = input.x;
+	setpoint.chosen_input.y = input.y;
+	setpoint.chosen_input.z = input.z;
+	setpoint.chosen_input.r = input.r;
+	// FIXME: what's that?
+	//setpoint.chosen_input.vx = (input.x - _manual_control_input[i].x) * dt_inv;
+	//setpoint.chosen_input.vy = (input.y - _manual_control_input[i].y) * dt_inv;
+	//setpoint.chosen_input.vz = (input.z - _manual_control_input[i].z) * dt_inv;
+	//setpoint.chosen_input.vr = (input.r - _manual_control_input[i].r) * dt_inv;
```

Comment:
This is just something on the side I was playing with initially for things like deciding if the sticks are moving. Now that all input sources are routed through this module we can track the update rates. This could be deleted for now.